### PR TITLE
Auto-generate dirac points in PorousFlowPolyLineSink

### LINKED ISSUE #15547: PorousFlowPolyLineSink auto point generation
## Reason
It would be nice if, when creating a PorousFlowPolyLineSink object, that we could automatically generate a point for each element the desired line sink/source passed through - instead of having to manually add points to a file.  This would help with parameter sweeps we are doing with some geothermal porous flow work (using the `falcon` app) in addition to just being generally convenient.

## Design
We could just use the existing ray-tracing utils to do this.  I have a prototype that seems to work - adding a point at the midpoint of each line segment chopped by the cell faces the line passes through.  Would @WilkAndy be interested/okay with this sort of functionality being added?

## Impact
Shouldn't affect existing APIs

### PR DESCRIPTION
**Pull request by rwcarlsen**
Fixes #15547.

### CODE REVIEW

**Comment by WilkAndy** [comment]
I can certainly review this, but maybe not today - we'll see.   I think you need to include some markdown documentation - probably a simple paragraph added to the existing doco.

File: modules/porous_flow/src/dirackernels/PorousFlowLineGeometry.C
Comment (loganharbour) [inline modules/porous_flow/src/dirackernels/PorousFlowLineGeometry.C:152]: Add an extra check to see if this point exists in the mesh up front so that you get a more meaningful out of bounds error?

**Review by loganharbour — CHANGES_REQUESTED** [review]

File: modules/porous_flow/src/dirackernels/PorousFlowLineGeometry.C
Comment (rwcarlsen) [reply modules/porous_flow/src/dirackernels/PorousFlowLineGeometry.C:152]: Not a bad idea.

**Comment by rwcarlsen** [comment]
yo yo - this is all green and ready to go.

File: modules/porous_flow/test/src/vectorpostprocessors/DiracPointsWriter.C
Comment (loganharbour) [inline modules/porous_flow/test/src/vectorpostprocessors/DiracPointsWriter.C:34]: `std::set<Point>` for sorting?

**Review by loganharbour — DISMISSED** [review]

File: modules/porous_flow/test/src/vectorpostprocessors/DiracPointsWriter.C
Comment (rwcarlsen) [reply modules/porous_flow/test/src/vectorpostprocessors/DiracPointsWriter.C:34]: Yes - the sorting matters for testing purposes - otherwise depending on the platform, you could get the points outputted in an inconsistent order - which actually caused test failures initially.

**Comment by rwcarlsen** [comment]
@WilkAndy - any comments here before we merge this?

**Comment by WilkAndy** [comment]
If you need this merged ASAP, go ahead.  I will only have time to review this in about 27 hours from now.

**Comment by rwcarlsen** [comment]
It's not urgently necessary - for our immediate need, we are just using patched moose internally.  I can wait a day.

**Comment by rwcarlsen** [comment]
It's not urgently necessary - for our immediate need, we are just using patched moose internally.  I can wait a day.

**Review by WilkAndy — DISMISSED** [review]
I wish i could have done this when i made the Dirac Kernels initially, thanks @rwcarlsen .    Without actually testing it myself, i think it looks good.   Two questions:

- have you changed any of the existing test files to see whether they still pass with this new scheme.  that'd be great to see.

- what happens if the line goes out of the mesh?   i'm pretty sure that with the old method MOOSE used to crash with a pretty uninformative error.

**Comment by rwcarlsen** [comment]
One of the tests was created by copying an existing test and gold file.  But it was a one-point-file case.  I haven't tried this with one of the multi-point files.

The problem with checking for the line going outside the mesh is that you already have use cases where you specify a line length in the third dimension of 2D problems - i.e. a simulated line source that really is like a point source.  If we add an error check for line end points out of the mesh then these cases won't work.

**Comment by WilkAndy** [comment]
> One of the tests was created by copying an existing test and gold file. But it was a one-point-file case. I haven't tried this with one of the multi-point files.
> 
> The problem with checking for the line going outside the mesh is that you already have use cases where you specify a line length in the third dimension of 2D problems - i.e. a simulated line source that really is like a point source. If we add an error check for line end points out of the mesh then these cases won't work.

Ha, i never thought about that 2D-3D thing and the line going out of the mesh.    Well, it doesn't matter a lot, providing the error message is understandable in the case the line does exit the mesh.

**Comment by rwcarlsen** [comment]
@WilkAndy - Are you saying to make the message an error?  We could make it a warning at least I suppose.  If we make it an error, than the test where I just used the same gold file from an existing test wouldn't be possible to run - or at least it would have to be rewritten - which sort of defeats the purpose of confirming that the auto-line code paths generate the same results as the manual points file.

**Comment by WilkAndy** [comment]
Some sort of message is necessary.  i would actually make it an error, just in case the user didn't notice the warning message, but i'll leave that up to you.

**Comment by rwcarlsen** [comment]
Okay - I think I've worked out an okay set of errors/tests now.  I added the line-exits-mesh error requested and modified the test to not trigger it incorrectly.  This should be good to go now.

**Comment by WilkAndy** [comment]
Great, thanks @rwcarlsen .   Let's merge this and see some people using it!

a

**Review by loganharbour — APPROVED** [review]

**Comment by rwcarlsen** [comment]
@milljm - I carefully investigated the app failure - it is in sockeye where the max ad dofs per elem is being violated.  This is unrelated to my PR which only modifies the porous flow module.  The sockeye failures also pre-existed in the next branch and other PRs.

**Pull request merged**
